### ISSUE TITLE
bug: VelStand PPO recovery does not learn after OOM-safe retraining

### ISSUE THREAD

**Issue by xyys2003** [unilabsim/microduck_rl_unilab#2]
## Owner summary

**目标**：报告 `microduck_rl_unilab` 中三个 PPO 任务的可复现训练参数和实际行为，定位 `MicroduckVelstandFlat` 在当前 MuJoCo 配置下无法完成 fall recovery 的问题。

**当前结论**：`MicroduckVelocityFlat` 在 MuJoCo 和 MJWarp 上均完成训练并表现出稳定行走；`MicroduckVelstandFlat + MuJoCo` 的训练循环完成，但摔倒恢复没有学会（`recovery_success=0`，`fallen_too_long` 仍非零）。

**交付边界**：本 issue 是训练行为 bug 的复现和证据记录，不在本 issue 中直接修改训练代码；不声称 VelStand 完整收敛；不把不同 seed/环境数/预算的 reward 差异当作受控后端 A/B 结论。

**建议类型**：bug / training behavior。

## Problem

`MicroduckVelstandFlat` 的定义目标是 walking + fall recovery，但在当前 MuJoCo owner 配置下出现了两个可复现的问题：

1. 原始 owner 配置 `num_envs=4096`、`max_iterations=20000` 在约 60.5 GB RAM 的主机上无法正常启动/持续运行：2048 初始化即 RAM OOM，1600 完整训练约第 29 轮 RAM OOM。
2. 为保持 `num_envs × iterations = 81,920,000` 不变，改用 `num_envs=1024`、`max_iterations=80000` 后训练循环可以完成，但 `recovery_success` 仍为 0，`fallen_too_long` 仍非零。
3. 从最终 checkpoint 的 40 秒可视化视频看，策略可以维持部分 walking/standing 行为，但跌倒后没有重新站立；因此 VelStand 的核心 recovery 目标没有学会。

这不是“结果无法判断”的文档问题，而是训练行为与 VelStand 任务目标不一致的 bug/失败复现；本 issue 同时提供三个任务的完整参数，便于区分 velocity 任务的正常结果和 VelStand 的失败结果。

## Expected behavior

- 在资源可承受的等价训练配置下，VelStand 应对其 recovery curriculum 产生非零的 `recovery_success` 事件，并在可视化 rollout 中展示跌倒后的重新站立。
- 如果 `4096 × 20000` 不是当前硬件的可运行配置，owner YAML 或 README 应明确资源要求，或者提供一个经过验证的安全默认配置，而不是让训练在初始化/早期迭代阶段以 RAM OOM 结束。

## Actual behavior

- `4096` 默认环境数无法在本机稳定运行；`1600` 也无法完成训练。
- `1024 × 80000` 可以完成训练，但 recovery event 没有出现，视频中没有观察到跌倒后重新站立。
- recovery shaping 并非始终关闭：`recovery_success` 在 step `28800`（iteration `1200`）后由 `0.0` curriculum 到 `10.0`，`upright_progress=5.0`、`height_progress=30.0` 一直启用；因此 `recovery_success=0` 不能解释成“奖励权重一直为零”。

## Evidence baseline

| 项目 | 本次环境 |
|---|---|
| Repository | `unilabsim/microduck_rl_unilab` |
| Code commit | `76cf9c0` (`main`) |
| Algorithm | PPO / `uni_rl.algos.rsl_rl_ppo:FinalObservationAwarePPO` |
| Host | AMD Ryzen 9 9950X3D2，16 cores；RAM 约 60.5 GB |
| GPU | NVIDIA GeForce RTX 4090，显存约 49,140 MiB |
| OS | Ubuntu 24.04；Linux `7.0.0-28-generic` |
| Runtime | `uv`，Python 3.13.x，TensorBoard |

run metadata 将工作树记录为 `dirty=true`；若要 bit-for-bit 复现，需要同时固定完整 `uv.lock`、依赖版本和干净工作树。本 issue 只记录实验代码基线，不把本地生成的临时 XML、lockfile 镜像源变更或二进制 checkpoint 提交到仓库。

## Reproduction matrix

公共 PPO/仿真参数：`num_steps_per_env=24`、`sim_dt=0.005 s`、`ctrl_dt=0.02 s`、最大 episode 时长 20 s；policy MLP `[512, 256, 128]`、ELU、learning rate `1e-3`、entropy coefficient `0.01`、5 epochs、4 mini-batches、`gamma=0.99`、`lambda=0.95`、adaptive KL schedule、desired KL `0.01`。训练阶段使用 `training.no_play=true`。

| Task / backend | `num_envs` | `max_iterations` | seed | `envs × iterations` | rollout env steps (`×24`) | 总耗时 | 吞吐 (env steps/s) | final reward | best reward | mean episode length | status |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---|
| `MicroduckVelocityFlat` + MuJoCo | 2048 | 500 | 1 | 1,024,000 | 24,576,000 | 8m19s | 51,891.1 | 97.19 | 117.30 | 931.43 | completed |
| `MicroduckVelocityFlat` + MJWarp | 4096 | 2000 | 42 | 8,192,000 | 196,608,000 | 9h25m27s | 5,799.7 | 109.90 | 130.78 | 915.04 | completed |
| `MicroduckVelstandFlat` + MuJoCo | 1024 | 80000 | 42 | 81,920,000 | 1,966,080,000 | 17h48m08s | 31,396.1 | 24.32 | 41.73 | 407.34 | completed |

`final reward` 是最后记录的 `Train/mean_reward`；`best reward` 是训练曲线最高记录，不是独立评估集分数。`completed` 只表示达到 `max_iterations`，不等价于所有 task sub-goal 都已学会。

### Velocity + MuJoCo

- owner config：`src/microduck_rl_unilab/conf/ppo/task/microduck_velocity_flat/mujoco.yaml`
- run：`2026-09-04_21-38-27_mujoco`
- checkpoint：`model_499.pt`
- result：final 97.19，best 117.30，mean episode length 931.43；视频显示基本稳定行走/直立。

### Velocity + MJWarp

- owner config：`src/microduck_rl_unilab/conf/ppo/task/microduck_velocity_flat/mjwarp.yaml`
- run：`2026-09-05_17-47-03_mjwarp`
- checkpoint：`model_1999.pt`
- result：final 109.90，best 130.78，mean episode length 915.04；视频显示基本稳定行走。
- 注意：不能仅凭 reward 高于 MuJoCo 就声称 MJWarp 策略质量或 sample efficiency 更好，因为 seed、环境数和总迭代预算不同。

### VelStand + MuJoCo

- owner config：`src/microduck_rl_unilab/conf/ppo/task/microduck_velstand_flat/mujoco.yaml`
- YAML 默认：`num_envs=4096`、`max_iterations=20000`。
- 最终运行：`num_envs=1024`、`max_iterations=80000`、seed `42`，保持 `num_envs × iterations = 81,920,000`，即与默认预算相同。
- run：`2026-09-08_19-31-27_mujoco`
- checkpoint：`model_79999.pt`
- result：final 24.32，best 41.73，mean episode length 407.34。
- recovery evidence：`recovery_success=0`；`fallen_too_long` 仍非零。因此准确结论是 walking/standing 部分有一定稳定性，但 fall recovery 尚未收敛。注意：`recovery_success` 的基础 weight 虽然为 `0.0`，但 curriculum 在 step `28800`（iteration `1200`）会将它提高到 `10.0`；`upright_progress=5.0` 和 `height_progress=30.0` 也一直提供 recovery shaping。因此这里的 `recovery_success=0` 指的是成功恢复事件没有被观察到，不是“整个恢复奖励始终关闭”。

## Memory limit investigation

| `num_envs` | observation |
|---:|---|
| 1024 | 单轮和完整训练均可运行；最终使用该值完成 80,000 iterations |
| 1536 | 单轮探测成功 |
| 1600 | 单轮探测成功，但完整训练约第 29 轮触发 RAM OOM；RSS 约 48.2 GB |
| 2048 | 初始化阶段 RAM OOM；内核记录 RSS 约 54.1 GB |

机器 RAM 约 60.5 GB，GPU 为 RTX 4090。当前证据支持：VelStand 的主要限制是 MuJoCo/BAM 批量状态和训练缓存造成的系统 RAM 压力，不是 GPU 显存，也不是环境数必须为 2 的幂次。1024 是当前安全且可完成训练的配置。

### Observed facts

1. 三个训练循环均正常结束；最终 VelStand run 没有 OOM。
2. 两个 velocity 任务在回报和视频上均达到基本稳定行走。
3. VelStand 的 `recovery_success` 没有形成稳定非零信号，`fallen_too_long` 仍然出现。该成功奖励并非全程关闭：它在 recovery curriculum 后期已打开，但成功事件仍为 0。
4. 更早的 `512 × 20000` VelStand run 最终 reward 约 25、best 约 44.88，与 `1024 × 80000` 的 24.32/41.73 同量级；仅增加训练预算没有解决恢复问题。

### Hypotheses to verify

- `recovery_success` 的姿态/高度/持续 fallen 阈值在当前接触和 BAM 动力学下可能难以触发。
- prone/crouch reset curriculum 可能提供了不足或过晚的 recovery exploration。
- `recovery_success`、`fallen_tax`、`com_upward_velocity` 的 reward curriculum 与 `fallen_too_long` 回收逻辑可能截断有效恢复轨迹。
- BAM 电压、扭矩变化惩罚、地面接触和 body-shell 碰撞可能使“翻身—支撑—起身”难以探索或不可达。

以上是待验证假设，不把其中任何一项当成已经证明的根因。后续应先记录 recovery state machine 的触发次数、阶段持续时间、prone/crouch reset 占比和 recovery reward 非零比例，再决定是否改奖励或初始化；不建议继续只增加训练时长。

## Reproduction commands

```bash
# Velocity + MuJoCo
uv run microduck-train --algo ppo --task microduck_velocity_flat --sim mujoco \
  training.no_play=true

# Velocity + MJWarp
uv run microduck-train --algo ppo --task microduck_velocity_flat --sim mjwarp \
  training.no_play=true

# VelStand + MuJoCo, fixed envs × iterations budget
uv run microduck-train --algo ppo --task microduck_velstand_flat --sim mujoco \
  algo.num_envs=1024 algo.max_iterations=80000 training.no_play=true
```

评估/录视频应在训练结束后单独运行，避免将渲染开销计入训练吞吐。

## Deliverables

- [ ] 将本文作为该仓库的公开实验记录保留。
- [ ] 附上三个与 run/checkpoint 对应的 40 秒 MP4 视频。
- [ ] 在 README 或后续实验索引中链接本 issue（如维护者希望长期公开这些结果）。
- [ ] 保留 VelStand“训练完成但恢复未收敛”的限制性表述，不将其改写为完整成功。

## Definition of done

- [ ] Issue 正文包含运行环境、commit、算法、seed、环境数、迭代数、rollout steps、耗时、吞吐、final/best reward、episode length。
- [ ] 三个视频在 GitHub Issue 中可以直接播放或通过附件链接下载，并标明 task/backend/run。
- [ ] 复现命令使用 `uv run`，且明确 VelStand 的 `1024 × 80000` 覆盖默认 `4096 × 20000`。
- [ ] 内存探测结果和 OOM 类型（system RAM，而非 GPU VRAM）有记录。
- [ ] 分析区分已观察事实与待验证假设；没有用不同 seed/预算的 reward 做无依据的后端优劣结论。
- [ ] 不提交二进制 checkpoint、MP4、临时 XML 或本地 lockfile 镜像源变更到代码树。

## Owner / dependencies / blockers

- **Proposed owner**：保持 unassigned，待仓库 maintainer 确认；本 issue 由实验执行者提供结果和附件。
- **Review owner**：仓库 maintainer / MicroDuck task owner。
- **Dependencies**：`unilab` / `unilab-rl` 依赖版本、MuJoCo/MJWarp runtime、RTX 4090 + 约 60 GB RAM 环境。
- **Current blocker**：仓库 Discussions 未开启，因此使用 Issue 记录；视频通过 Issue 附件上传。

## Artifacts

待作为 Issue 附件上传：

1. `MicroduckVelocityFlat + MuJoCo` — `play_video.mp4`，40 s，748,721 bytes。
2. `MicroduckVelocityFlat + MJWarp` — `play_video.mp4`，40 s，4,887,702 bytes。
3. `MicroduckVelstandFlat + MuJoCo` — `play_video.mp4`，40 s，633,227 bytes。

本地 checkpoint/run 目录（用于生成以上视频和本地核对）：

- `logs/rsl_rl_ppo/MicroduckVelocityFlat/2026-09-04_21-38-27_mujoco/`
- `logs/rsl_rl_ppo/MicroduckVelocityFlat/2026-09-05_17-47-03_mjwarp/`
- `logs/rsl_rl_ppo/MicroduckVelstandFlat/2026-09-08_19-31-27_mujoco/`

**Comment by xyys2003** [comment]
https://github.com/user-attachments/assets/57295282-f5be-4677-974d-f4c7d6adb37e

https://github.com/user-attachments/assets/3304c5f6-904d-4b88-a645-e4e443ae7260

https://github.com/user-attachments/assets/bb8a48b9-2a35-4a2e-b430-bd6c80f330d1

**Comment by xyys2003** [comment]
[microduck_tensorboard_exports.zip](https://github.com/user-attachments/files/31995333/microduck_tensorboard_exports.zip)
